client/dialogs/dialog-reducer.ts

```diff
@@ -1,23 +1,28 @@
-import { Map, Record } from 'immutable'
-import { keyedReducer } from '../reducers/keyed-reducer'
+import { Immutable } from 'immer'
+import { immerKeyedReducer } from '../reducers/keyed-reducer'
 import { DialogType } from './dialog-type'
 
-export class DialogState extends Record({
-  isDialogOpened: false,
-  dialogType: null as DialogType | null,
-  initData: Map(),
-}) {}
+export interface DialogState {
+  type: DialogType
+  initData?: Record<string, unknown>
+}
 
-export default keyedReducer(new DialogState(), {
+export interface DialogHistoryState {
+  history: DialogState[]
+}
+
+const DEFAULT_DIALOG_HISTORY_STATE: Immutable<DialogHistoryState> = {
+  history: [],
+}
+
+export default immerKeyedReducer(DEFAULT_DIALOG_HISTORY_STATE, {
   ['@dialogs/open'](state, action) {
-    return new DialogState({
-      isDialogOpened: true,
-      dialogType: action.payload.dialogType,
-      initData: Map(Object.entries(action.payload.initData)),
-    })
+    const { type, initData } = action.payload
+
+    state.history.push({ type, initData })
   },
 
   ['@dialogs/close'](state, action) {
-    return new DialogState()
+    state.history.pop()
```

Comment:
I feel like if we're going to allow stacking, the `close` calls should probably be more specific to a particular dialog type, or this gets pretty dangerous. I'd add a dialog type parameter to this + allow them to specify `all` as well. Specifying a particular type will pop the stack until it removes a dialog of that type (or do nothing if no dialog of that type was open?). Specifying `all` clears the stack completely.